### ISSUE TITLE
[Question] Tutorial 11, Variational dequantization

### ISSUE THREAD

**Issue by mjack3** [phlippe/uvadlc_notebooks#115]
**Tutorial 11**

Hello, I hope you are well. I am studying for a Ph.D. in normalizing flow, and this tutorial is a fantastic help. However, I have five questions I would appreciate if you could answer.

I have doubts about cell number 10:
```
img = (z / 255.0) * 2 - 1 # We condition the flows on x, i.e. the original image
# Prior of u is a uniform distribution as before
# As most flow transformations are defined on [-infinity,+infinity], we apply an inverse sigmoid first.
deq_noise = torch.rand_like(z).detach()
deq_noise, ldj = self.sigmoid(deq_noise, ldj, reverse=True)
```

I wonder:
1. why do you want `img` in the range [-1, 1]
2. why do you want the `deq_noise` in the range [-inf, +inf]
3. Is that because you apply the inverse of the sigmoid to the images throughout the other flows after de-quantizing?
4. Isn't it enough to use the deq_noise without conditioning the flow to the image? I don't get the idea of conditioning the noise to the transformed image in the range [-1, 1].

Extra:

In my case, I am familiar with the FrEIA framework, and they implement the conditional INN in this way. They first split the input tensor `x`. Then, they add the condition `c` to the half and pass it through the subnet.

```
x1, x2 = torch.split(x[0], self.splits, dim=1)
if self.conditional:
          x1c = torch.cat([x1, *c], 1)
          a1 = self.subnet(x1c)
```

You first concatenate the condition, ingest the subnet, and split the vector.

```
if orig_img is None:
          nn_out = self.network(z_in)
else:
          nn_out = self.network(torch.cat([z_in, orig_img], dim=1))
          s, t = nn_out.chunk(2, dim=1)
```

Is there something I'm missing?

Thanks for your time!

**Comment by cviviers** [comment]
@mjack3 in your second question, the mask gets applied to z_in, effectively splitting it. The masked z_in and condtion gets passed through the network to predict s,t which then gets applied to the z_in multiplied by the inverse of the mask, i.e. the "other" half. While the two implementations are not exactly the same, I think they achieve the same thing. 

Also hope @phlippe can shed some light on this and your first questions

**Comment by phlippe** [comment]
Hi @mjack3, thanks for your questions, and thanks @cviviers for already answering the second question! Let me try to give you some more intuition below.

> 1. why do you want img in the range [-1, 1]

This image will be the input to a neural network. As with any neural network, you want the input to be ideally zero-centered and have a standard deviation of 1 (rest of the network is designed for that, e.g. see initialization - Tutorial 4). A cheap, approximate way of getting this with images is to normalize them in the range [-1, 1].

> 2. why do you want the deq_noise in the range [-inf, +inf]

We will apply normalizing flows on this deq_noise, which are defined for an input range of [-inf, +inf] and output range of [-inf, +inf]. If we would put in a range of [0, 1], we are not guaranteed to get actually outputs in the range [0, 1], which means the flow may not be invertible in your expected range. Going to the [-inf, +inf] makes your life easier, since you can freely choose from your flow operations.

> 3. Is that because you apply the inverse of the sigmoid to the images throughout the other flows after de-quantizing?

Before we dequantize, we will map this noise back to [0, 1] (see the sigmoid 2 lines later). We want the flows to operate on [-inf, +inf], and an easy way to invertibly transition between [-inf, +inf] and [0, 1] is a sigmoid (or its inverse for reverse).

> 4. Isn't it enough to use the deq_noise without conditioning the flow to the image? I don't get the idea of conditioning the noise to the transformed image in the range [-1, 1].

Conditioning on the image allows your model to flexible adjust the noise distribution for each individual image, which represents a sub-distribution in your global marginal distribution (basically $p(z|x)$ vs $p(x)$ with $z$ being the latent variables and $x$ the image). If you have a completely black image, you may want to use different noise distributions on these black pixel versus a very bright image with rare black pixels. Overall, you can make the life of the model much easier by that.

> Extra

As @cviviers already correctly mentioned, they are basically doing the same thing. The masking strategy, commonly used in flows, simply supports any form of mask, while the splitting approach requires a certain structure in x such that you can split. 

Hope this helps, let me know if you have any follow-up questions!

**Comment by mjack3** [comment]
Great! Thanks!
